<a href="https://colab.research.google.com/github/yewh86/utmmachinelearning/blob/master/Quiz2_Felicia_Yew.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()


Saving heart.csv to heart.csv


In [0]:
import keras
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.optimizers import SGD
import numpy as np
from sklearn.model_selection import train_test_split
from keras.utils import to_categorical
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [16]:
data = pd.read_csv('heart.csv')

type(data)

data.shape

(303, 14)

In [18]:
data.head(5)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [19]:
data.sample(5)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
66   51    1   2       100   222    0        1      143      1      1.2   
81   45    1   1       128   308    0        0      170      0      0.0   
16   58    0   2       120   340    0        1      172      0      0.0   
37   54    1   2       150   232    0        0      165      0      1.6   
36   54    0   2       135   304    1        1      170      0      0.0   

    slope  ca  thal  target  
66      1   0     2       1  
81      2   0     2       1  
16      2   0     2       1  
37      2   0     3       1  
36      2   0     2       1

In [20]:
data.describe()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
age         303 non-null int64
sex         303 non-null int64
cp          303 non-null int64
trestbps    303 non-null int64
chol        303 non-null int64
fbs         303 non-null int64
restecg     303 non-null int64
thalach     303 non-null int64
exang       303 non-null int64
oldpeak     303 non-null float64
slope       303 non-null int64
ca          303 non-null int64
thal        303 non-null int64
target      303 non-null int64
dtypes: float64(1), int64(13)
memory usage: 33.2 KB


In [0]:

predictors = data.drop("target", axis=1)
target = data["target"]

X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size = 0.30, random_state=42)


In [65]:
model = Sequential()
model.add(Dense(11, activation='relu', input_dim=13))
model.add(Dropout(0.5))
model.add(Dense(6, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
model.summary()


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_59 (Dense)             (None, 11)                154       
_________________________________________________________________
dropout_39 (Dropout)         (None, 11)                0         
_________________________________________________________________
dense_60 (Dense)             (None, 6)                 72        
_________________________________________________________________
dropout_40 (Dropout)         (None, 6)                 0         
_________________________________________________________________
dense_61 (Dense)             (None, 1)                 7         
Total params: 233
Trainable params: 233
Non-trainable params: 0
_________________________________________________________________


In [75]:
model.fit(X_train, y_train, validation_split=0.10, epochs = 20, batch_size=128)


Train on 190 samples, validate on 22 samples
Epoch 1/20
190/190 [==============================] - 0s 53us/step - loss: 0.6878 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 2/20
190/190 [==============================] - 0s 47us/step - loss: 0.6950 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 3/20
190/190 [==============================] - 0s 42us/step - loss: 0.6662 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 4/20
190/190 [==============================] - 0s 35us/step - loss: 0.6935 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 5/20
190/190 [==============================] - 0s 37us/step - loss: 0.6893 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 6/20
190/190 [==============================] - 0s 48us/step - loss: 0.6844 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 7/20
190/190 [==============================] - 0s 46us/step - loss: 0.6859 - acc: 0.5368 - val_loss: 0.6844 - val_acc: 0.5909
Epoch 8/20
190/190 [====

In [73]:
score = model.evaluate(X_test, y_test, batch_size=128)

91/91 [==============================] - 0s 20us/step


2. What type of problem are we trying to solve here? Classification or regression? 

The problem is to distinguish presence of cardiovascular disease and to identify trend to predict certain cardiovascular events or indications of heart health.  Classification. 


---



3. You have decided you want to create a neural network model to perform the classification/regression for that problem. 
a. Use train_test_split function to split the data into training 70%, test 30%
b. Design a neural network model with any architecture. Try to be unique here.
c. Use Keras codes to create the model
d. Compile and fit the model
e. What is the accuracy after 20 epochs?Accuracy is 59.09%


---


4. What is underfitting and overfitting?

Underfitting is referring to a model that can neither model the training data nor generalize to new data. It is not suitable model with poor performance on the training data. 

Overfitting refers to a model that resulting of training data too well. Overfitting happens when a model learns the detail and noise in the training data to the extent that it negatively impacts the performance of the model on new data. **bold text**


---


5. Explain how you can reduce the overfitting of the model.
Include parameters or techniques which limit and constrain how much detail the model learns. 
